리뷰 스터디용 PR

RaZel713 opened the pull request:
# 다들 이번 미션 고생하셨습니다!

<PR 업로드 위치>
내 저장소 PR에서 메인 브랜치 <- 진행 브랜치 로

<리뷰 기한>
금요일 오후 6시 ~ 다음주 월요일 리뷰 모임 전

<리뷰 방식>
- 최소 3개 ~ 갯수 제한 없이 (적당히!)
- 읽어봤을 때 의도를 알기 어려웠던 부분, 이해가 안되는 부분 위주로 해주시면 좋습니다!
- 상수처리, 추상화 레벨 맞추기, 네이밍 등 간단한 점들은 언제든 오케이!
- 좋은점 배운점 작성도 좋아요!

RaZel713 committed:
[1 - 3단계 방탈출 사용자 예약] 라젤(문선영) 미션 제출합니다. (#169)

* init: 프로젝트 초기 설정

* docs: API 명세 작성

* refactor: Admin 컨트롤러에서 정적 콘텐츠 관리

* feat: theme 테이블 추가 및 관계 설정

* feat: 테마 관리 페이지 응답 메서드 추가

* feat: 테마 도메인 추가

* feat: 테마 DAO 추가

* feat: 테마 조회 기능 추가

* refactor: DTO 패키지 controller로 이동

* feat: ThemeDAO 저장 기능 추가

* feat: 테마 추가 기능 추가

* refactor: 컨트롤러 DTO 패키지 분리

- 요청과 응답 패키지로 분리

* refactor: 컨트롤러 클래스 네이밍 변경 및 패키지 분리

* refactor: 요청 DTO의 역할에 맞는 네이밍 변경

* refactor: CRUD 순서로 메서드 정렬

* refactor: 컨트롤러의 요청 DTO와 서비스 처리 DTO 분리

* test: e2e 테스트 삭제

* feat: 테마 삭제 기능 추가

* fix: URL 경로명 오타 수정

* refactor: 스키마 변경

* feat: 초기 데이터 삽입 sql 추가

* feat: 예약 시, 테마 설정 기능 추가

* feat: 커스텀 예외 처리 기능 추가

* fix: 쿼리 오류 수정

* feat: 커스텀 예외 적용

* feat: 예약, 예약 시간 도메인 생성자에 빈값 검증 추가

* refactor: 미사용 메서드 삭제

* feat: 데이터 추가, 삭제 시 적합한 http 상태 코드 반환

- 데이터 추가 성공 시, 201 CREATED 반환
- 데이터 삭제 성공 시, 204 NO CONTENT 반환
- 데이터 삭제 실패 시, 404 NOT FOUND 반환

* refactor: 예외 처리 로직을 controller에서 service로 이동

* feat: 중복되는 테마 저장 검증 기능 추가

* feat: 예약 시 시간 검증 기능 추가

- 과거 시점 예약 시 예외 반환

* feat: 사용 중인 데이터를 삭제 시도 시 예외 처리 기능 추가

* refactor: controller에 @ResponseStatus 적용

* refactor: controller 네이밍 변경

* test: 임시 데이터 값 수정

* feat: 사용자 예약 페이지 응답 기능 추가

* feat: 예약 가능 시간 조회 기능 추가

* test: 임시 데이터 값 추가

* refactor: 메서드 명 오타 수정

* feat: 인기 테마 목록 조회 기능 추가

* refactor: 코드 순서 정렬

* refactor: 코드 컨벤션 수정

- 클래스 선언부와 필드 선언부 사이에 공백 라인 추가

* test: http 요청 형식 수정

* refactor: final 키워드 적용

* refactor: 테마 도메인 네이밍 변경

* refactor: 커스텀 예외 오버라이드 메서드 삭제

* test: 임시 데이터 값 수정

* test: 기존 테스트 수정

* test: 예약 서비스 기능 테스트 추가

* test: 예약 시간 서비스 기능 테스트 추가

* docs: readme 작성

* refactor: 테마 도메인 네이밍 변경

* refactor: 클래스명 오타 수정

* test: 테마 서비스 기능 테스트 추가

* refactor: 클래스명 오타 수정

* test: 테마 서비스 기능 테스트 추가

* refactor: 서비스 리턴 타입 수정

* refactor: 불필요한 todo 주석 삭제

* refactor: 메서드 네이밍 변경

* refactor: 불필요한 메서드 제거

* refactor: 서비스 메서드 분리

- findById를 생성해 추상화 적용

* test: 미사용 테스트 삭제

* test: 테마 랭킹 조회 http 추가

* refactor: 상수의 접근 제한자를 private 으로 변경

* refactor: 매직 넘버 상수 처리

* feat: 인기 테마 날짜 범위 도메인 추가

- 인기 테마 선정 기준일과 선정 기간 일 수를 받아 인기 테마 선정 날짜 범위를 계산하는 도메인 클래스를 추가한다.

* refactor: 과거 시점 검사 로직을 Reservation 도메인으로 이동

- 서비스에서 처리하던 예약 시점 유효성 검사를 도메인 내부 메서드로 이동하여 도메인이 자신의 상태를 스스로 판단할 수 있도록 한다.

* refactor: 예약 가능 시간 응답 생성 로직을 변수 추출하여 가독성 향상

* refactor: 예약 가능 시간 응답 생성 메서드 네이밍 구체화

- AvailableReservationTimeResponse의 정적 팩토리 메서드명을 from → fromReservationTimeWithBookingStatus로 변경한다.
- 메서드의 인자가 예약 시간과 예약 여부를 포함하므로, 의미를 명확히 드러내도록 개선한다.

* refactor: 인기 테마 개수 제한을 쿼리에서 제거하고 서비스 로직으로 이동

- 쿼리에 비즈니스 요구사항이 섞여 있으면 요구사항 변경 시 쿼리 수정이 필요해 유지 보수가 어렵다.
- 따라서 쿼리에서 LIMIT 10을 제거하고, Java 코드에서 상위 10개만 추려내도록 변경한다.

* refactor: 예약 도메인 정렬 순서 변경

* test: 예약 도메인 테스트 추가

* refactor: 예약 시간 도메인 정렬 순서 변경

* test: 예약 시간 도메인 테스트 추가

* test: 방 테마 도메인 테스트 추가

* test: 인기 테마 선택 기준 도메인 테스트 추가

* refactor: 서비스가 도메인을 반환하고 컨트롤러에서 DTO로 변환하도록 수정

- 서비스 계층에서 컨트롤러의 Response DTO에 대한 의존 제거한다
- 컨트롤러에서 도메인을 받아 DTO로 변환하는 책임을 수행한다
- 계층 간 의존 방향을 컨트롤러 → 서비스 라는 것을 명확히 한다

* test: 기능별 http 작성 및 패키지 이동

* test: 예약 DAO 테스트 추가

* test: 예약 시간 DAO 테스트 추가

* test: 방 테마 DAO 테스트 추가

* refactor: 클래스 네이밍 수정

* test: 변수명 변경

* feat: 사용 중인 시간을 삭제할 경우 발생하는 예외 추가 및 비즈니스 규칙 위반 예외 네이밍 변경

* refactor: '@ResponseStatus' 애너테이션을 사용하여 메서드 별 HTTP 상태 코드 명시

* refactor: 줄바꿈 변경

* refactor: 유효성 검사 메서드 추출로 메서드 간 추상화 수준 일치

* refactor: 메서드 네이밍 변경

* refactor: 도메인의 생성자 파라미터 유효성 검증 메서드 네이밍 변경

* feat: 인기 테마 선정 기준 도메인의 생성자 파라미터 유효성 검증 메서드 추가

* refactor: final 키워드 추가

* feat: 예약자 명 길이 검증 메서드 추가

* feat: 테마 명, 테마 설명, 썸네일 주소 길이 검증 메서드 추가

* test: assertAll 적용

* test: 데이터 베이스 연결 테스트 추가

* test: 어드민 페이지 컨트롤러 테스트 추가

* test: 예약 페이지 컨트롤러 테스트 추가

* test: 예약 시간 테스트 데이터 변경

* test: 예약 테마, 예약 테스트 데이터 변경

* test: 예약 컨트롤러 테스트 추가

* test: 예약 시간 컨트롤러 테스트 추가

* test: 방 테마 컨트롤러 테스트 추가

* feat: 정의된 예외를 제외한 모든 오류 발생 시 500 Internal Server Error 상태 코드 반환

- 비즈니스 로직에서 정의한 예외를 제외한 예기치 못한 오류 발생 시 500 상태 코드 반환한다

* refactor: 도메인 패키지 분리

- 예약, 예약 시간, 테마 관련 클래스를 도메인 패키지 하위에 분리한다

* test: 테스트 데이터 변경에 따른 HTTP 파일 수정

* refactor: 인기 테마 조회 시 상위 개수 제한을 파라미터로 받도록 변경

* test: 테스트 메서드 명 변경

* refactor: 컨트롤러와 서비스 DTO 네이밍 변경

* refactor: 예약, 예약 시간 도메인의 과거 시점 판단 메서드 명 변경

- 시점을 판별하여 boolean 값을 반환하므로 is로 시작하는 메서드 명으로 변경한다

* refactor: 서비스 DTO 네이밍 변경

* refactor: service에서 insert 이후 반환된 id의 값을 넣은 객체를 반환

* refactor: 컨트롤러를 rest와 view 패키지로 분리

---------

Co-authored-by: jhan0121 <jhan0121@gmail.com>
Co-authored-by: jhan0121 <56645802+jhan0121@users.noreply.github.com>

RaZel713 committed:
refactor: 누락된 접근 제한자 구현

RaZel713 committed:
refactor: CreateReservationRequest import 추가

RaZel713 committed:
docs: README 정리

RaZel713 committed:
feat: 로그인 페이지 컨트롤러 구현

RaZel713 committed:
feat: 멤버 테이블 스키마 정의

RaZel713 committed:
feat: 로그인 기능 구현

RaZel713 committed:
feat: 인증 정보 조회 기능 구현

RaZel713 committed:
feat: 로그인 기능 구현에 따른 html 추가

RaZel713 committed:
test: 로그인 기능 http 파일 구현

RaZel713 committed:
feat: 멤버 데이터 추가

RaZel713 committed:
refactor: 패키지 구조 변경

RaZel713 committed:
feat: 예약 테이블 스키마 정의

- 멤버 테이블을 외래키로 추가

RaZel713 committed:
refactor: DTO 네이밍 변경

RaZel713 committed:
feat: 멤버 도메인 유효성 검사 구현

RaZel713 committed:
feat: 예약 생성 시, 로그인한 사용자 정보를 활용

RaZel713 committed:
refactor: 예약 데이터 수정

RaZel713 committed:
feat: 관리자 예약 생성 시, 유저를 조회해 선택 후 생성

RaZel713 committed:
refactor: 패키지 구조 변경

RaZel713 committed:
refactor: 멤버 도메인 네이밍 변경

RaZel713 committed:
feat: 접근 권한 제어 기능 구현

RaZel713 committed:
feat: 예약 목록 검색 기능 구현

RaZel713 committed:
test: 테스트 패키지 구조 변경

- 프로덕션 코드의 패키지 구조와 일치하도록 테스트 패키지를 정리한다

RaZel713 committed:
refactor: 도메인 명과 데이터 테이블 명이 일치하도록 변경

RaZel713 committed:
test: 로그인 구현에 따른 예약 테스트 수정

RaZel713 committed:
test: 접근 권한 제어 기능 구현에 따른 도메인 페이지 테스트 수정

RaZel713 committed:
refactor: 메서드 네이밍 변경

RaZel713 committed:
refactor: 응답 DTO에서 비밀 번호 삭제

RaZel713 committed:
test: 로그인 페이지 컨트롤러 테스트 추가

RaZel713 committed:
test: 로그인 컨트롤러 테스트 추가

RaZel713 committed:
refactor: 응답 DTO 수정

RaZel713 committed:
fix: 어드민 페이지에서 예약 추가 오류 수정

KoSeonJe commented on src/http/Login.http:
http 파일 정말 잘쓰셨네요!! 최고다 bb

RaZel713 replied:
감사합니다!!

RaZel713 committed:
feat: 멤버 도메인에 이메일 형식 검증 로직 구현

RaZel713 committed:
refactor: final 키워드 추가

RaZel713 committed:
feat: 인증 만료 기한 단위를 주석으로 명시

KoSeonJe commented on src/main/java/roomescape/auth/CheckLoginInterceptor.java:
```diff
@@ -0,0 +1,35 @@
+        if (member == null || !member.getRole().equals("Admin")) {
+            response.setStatus(401);
+            return false;
+        }
```
예외처리 해도 좋을 것 같아요!
Exceptionhandler가 controller단의 예외를 핸들링 해주는데,
Interceptor의 prehandle메서드는 controller에 도달하기 전 실행되는데, 이 예외를 캐치할 수 있을까? 고민했었습니다.

근데, interceptor의 prehandler 예외를 특별히! 잡아서 핸들링 해준다고합니다. 자세한 동작 원리는 잘 몰라요! 궁금하면 같이 코드 까보는거로!

401과 403의 차이를 고민해보면 좋을 것도 같아요!

RaZel713 committed:
refactor: 메서드 명 수정

- 인증 서비스의 'findMember' 메서드 명을 'findMemberByToken'로 변경하여 토큰을 기준으로 멤버를 찾는 메서드 임을 명시한다.

KoSeonJe commented on src/main/java/roomescape/auth/service/dto/CreateTokenServiceRequest.java:
```diff
@@ -0,0 +1,10 @@
+    public static CreateTokenServiceRequest from(final LoginRequest request) {
+        return new CreateTokenServiceRequest(request.email(), request.password());
+    }
+}
```
이렇게되면 service 레이어 내에 존재하는 service request가 controller의 loginRequest를 의존하게 되어서, 역방향 의존성이 생기는 것 같아요!

LoginRequest내부에서 CreateTokenServiceRequest로 변환시켜주는 건 어때요?

RaZel713 replied:
피드백 반영하였습니다!

KoSeonJe commented on src/main/java/roomescape/auth/controller/dto/MemberNameResponse.java:
```diff
@@ -0,0 +1,10 @@
+
+import roomescape.member.domain.Member;
+
+public record MemberNameResponse(String name) {
```
클래스명이 확장성이 조금 떨어지는 것 같아요ㅠ
해당 API가 어떤 정보를 제공하고자 하는지를 포함하는 네이밍으로 수정하면 좋을 것 같아요!
인증정보를 담은 응답 객체라는 표현? ㅎㅎ

KoSeonJe commented on src/main/java/roomescape/auth/controller/LoginController.java:
```diff
@@ -0,0 +1,45 @@
+        this.authService = authService;
+    }
+
+    @PostMapping("/login")
```
로그인은 왜 POST 메서드일까 고민해보신적 있나요!?

KoSeonJe commented on src/main/java/roomescape/auth/service/AuthService.java:
```diff
@@ -0,0 +1,76 @@
+                .orElseThrow(() -> new AuthorizationException("존재하지 않는 이메일 입니다"));
+    }
+
+    public ResponseCookie createCookie(String accessToken) {
+        return ResponseCookie.from("token", accessToken)
+                .httpOnly(true)
+                .secure(false)
+                .path("/")
+                .sameSite("Lax")
+                .maxAge(60 * 60)
+                .build();
+    }
+
+    public String extractTokenFromCookie(final Cookie[] cookies) {
+        validateCookiesExisted(cookies);
+
+        return Arrays.stream(cookies)
+                .filter(cookie -> "token".equals(cookie.getName()))
+                .map(Cookie::getValue)
+                .findFirst()
+                .orElse("");
+    }
```
쿠키 관련 로직을 비즈니스 로직이라고 생각하시나요!?

KoSeonJe commented on src/main/java/roomescape/member/dao/JdbcMemberDao.java:
```diff
@@ -0,0 +1,50 @@
+        return jdbcTemplate.query(query, MEMBER_ROW_MAPPER);
+    }
+
+    @Override
+    public Optional<Member> findById(final long id) {
+        final String query = "SELECT * FROM member WHERE id = ?";
+        final List<Member> members = jdbcTemplate.query(query, MEMBER_ROW_MAPPER, id);
+        return members.stream()
+                .findFirst();
+    }
```
오 방법 좋은 것 같아요

KoSeonJe commented on src/main/java/roomescape/reservation/controller/ReservationController.java:
```diff
@@ -0,0 +1,76 @@
+        return ReservationResponse.from(savedReservation);
+    }
+
+    @PostMapping("admin/reservations")
```
```suggestion
    @PostMapping("/admin/reservations")
```
ㅎㅎ

KoSeonJe replied:
어드민용 컨트롤러를 생성하는 것에 대해 어떻게 생각하시나요?

KoSeonJe reviewed (commented):
고생하셨어요!
코멘트는 한번 슥 훑어주시고 생각해주시면 좋을 것 같아요!

RaZel713 committed:
refactor: admin 검증을 interceptor에서 service로 위임

RaZel713 committed:
refactor: AuthService가 구현체가 아닌 인터페이스에 의존하도록 리팩터링

- jjwt 외부 라이브러리에 대한 강한 결합을 줄이기 위해 AuthTokenProvider 인터페이스를 도입한다.
- JwtTokenProvider는 AuthTokenProvider 구현체로 변경한다.

RaZel713 committed:
refactor: 토큰 생성과 추출 책임을 AuthTokenProvider와 AuthTokenExtractor로 분리

RaZel713 committed:
refactor: AuthTokenProvider와 AuthTokenExtractor의 메서드 이름을 역할에 맞게 변경

RaZel713 committed:
refactor: @Value를 생성자 파라미터에 적용하여 의존성 주입

- secretKey와 validityInMilliseconds 값을 생성자 파라미터에서 주입받아 불변 필드로 초기화한다.

RaZel713 committed:
refactor: 불필요한 issuedAt 클레임 제거

RaZel713 committed:
refactor: 멤버 역할을 문자열 대신 Role enum으로 표현

RaZel713 committed:
refactor: 유틸 클래스로 쿠키 생성 및 추출 로직 분리

RaZel713 committed:
refactor: 컨트롤러 필드명 변경

RaZel713 committed:
refactor: 멤버 도메인의 이름, 이메일, 비밀 번호 길이 제한 변경

RaZel713 committed:
test: 멤버 도메인 테스트 구현

RaZel713 committed:
feat: 회원 가입 기능 구현

RaZel713 committed:
refactor: 누락된 접근 제한자 추가

RaZel713 committed:
test: 테스트 데이터 형식 수정

RaZel713 committed:
feat: 로그 아웃 기능 구현

RaZel713 committed:
docs: README 수정

RaZel713 committed:
test: 회원 가입 페이지 테스트 추가

RaZel713 committed:
test: 멤버 컨트롤러 테스트 추가

RaZel713 committed:
test: 멤버 서비스 테스트 추가

RaZel713 committed:
test: 멤버 DAO 테스트 추가

RaZel713 committed:
refactor: interceptor 패키지 생성

RaZel713 committed:
test: JwtTokenExtractor 테스트 추가

RaZel713 committed:
test: JwtTokenProvider 테스트 추가

RaZel713 committed:
test: CookieHandler 테스트 추가

RaZel713 committed:
test: 로그아웃 테스트 추가

RaZel713 committed:
test: 인증 서비스 테스트 추가

kwonkeonhyeong commented on src/main/java/roomescape/auth/interceptor/CheckLoginInterceptor.java:
```diff
@@ -0,0 +1,33 @@
+                             @NonNull Object handler) {
+
+        final Cookie[] cookies = request.getCookies();
+        final String token = CookieHandler.extractTokenFromCookies(cookies);
```
CookieHandler 라는 Util 클래스를 사용하고 있네요!
CookieHandler 내부를 확인해 보니 createCookieFromToken를 이용하여 token 을 추출하고 있는데, 
token 의 키 값이 변경 된다거나 추출 조건이 변경되면 CookieHandler를 사용하는 모든 곳이 이러한 변경에 유연하게 대처하기가 힘들 것 같다는 생각이 들었습니다!
이러한 변경 사항에 유연하게 대처할 수 있는 방법은 없을까요? 라젤의 생각이 궁금합니다!

+ AuthTokenExtractor에서 제공하는 Extract 기능과 CookieHandler 내부에 구현한 Extract기능을 따로 분리해서 구현한 이유도 궁금합니다!

kwonkeonhyeong commented on src/main/java/roomescape/auth/resolver/LoginMemberArgumentResolver.java:
```diff
@@ -0,0 +1,39 @@
+
+    @Override
+    public boolean supportsParameter(MethodParameter parameter) {
+        return parameter.hasParameterAnnotation(AuthenticationPrincipal.class);
```
HandlerMethodArgumentResolver 를 애노테이션 기반으로 구현하는 것과 클래스 기반으로 구현하는 것의 차이가 어떤 것이 있을까요?

kwonkeonhyeong commented on src/main/java/roomescape/auth/controller/LoginController.java:
```diff
@@ -0,0 +1,58 @@
+        response.addHeader(HttpHeaders.SET_COOKIE, cookie.toString());
+    }
+
+    @PostMapping("/logout")
```
ResponseEntity를 사용하는 곳과 사용하지 않는 곳을 어떤 기준으로 나누는지 라젤의 기준이 궁금합니다! (순수 궁금증입니다~)

saera-yook commented on src/main/java/roomescape/reservation/controller/ReservationController.java:
```diff
@@ -0,0 +1,76 @@
+    @ResponseStatus(HttpStatus.CREATED)
+    public ReservationResponse addReservation(
+            @RequestBody CreateUserReservationRequest request,
+            @AuthenticationPrincipal Member member) {
```
@AuthenticationPrincipal을 통해 도메인 객체인 Member를 직접 주입받는 구조를 선택하신 이유가 궁금합니다 :)
저는 인증 정보로부터 ID만 추출해 Service에서 도메인을 조회하는 구조로 분리하면 보안 측면에서 더 안전하고 계층 간 책임도 명확해지지 않을까 생각했는데 라젤이 선택한 방식의 장점도 알고 싶어서요ㅎㅎ

saera-yook commented on src/main/java/roomescape/auth/resolver/AuthenticationPrincipal.java:
```diff
@@ -0,0 +1,11 @@
+
+@Target(ElementType.PARAMETER)
+@Retention(RetentionPolicy.RUNTIME)
+public @interface AuthenticationPrincipal {
```
Spring Security에서 SecurityContextHolder.getContext().getAuthentication().getPrincipal()과 같은 인증 객체를 주입할 때 사용하는 @AuthenticationPrincipal 어노테이션이 이미 있다고 하더라구요...!
혼동이 없도록 다른 이름을 붙여보는 것은 어떨까요?

saera-yook commented on src/main/java/roomescape/auth/jwt/AuthTokenExtractor.java:
```diff
@@ -0,0 +1,8 @@
+package roomescape.auth.jwt;
+
+public interface AuthTokenExtractor {
```
AuthTokenExtractor와 TokenProvider를 인터페이스로 나눈것이 좋은 것 같아요!
저는 생각 못했는데 배워갑니다 :)

saera-yook reviewed (commented):
라젤 안녕하세요~

RaZel713 committed:
refactor: 로그인 체크 응답 DTO 네이밍 변경

RaZel713 committed:
refactor: 컨트롤러 요청 DTO에서 서비스 요청 DTO를 생성할 수 있도록 변경

RaZel713 committed:
refactor: 정적 팩토리 메서드를 활용

RaZel713 committed:
test: USER 토큰으로 접속 시, 권한 판별 테스트 추가

RaZel713 committed:
test: 불필요한 setUp 메서드 제거

RaZel713 committed:
feat: 인증 예외를 추가해 상황에 맞게 인증, 인가 예외를 분리

RaZel713 committed:
refactor: schema의 member 테이블 생성시, role을 enum이 아닌, varchar로 생성

RaZel713 committed:
refactor: 불필요한 메서드 래핑 제거

RaZel713 committed:
refactor: 어드민 페이지 유저 계정 접속 테스트 추가

RaZel713 committed:
refactor: 주석 수정

RaZel713 committed:
refactor: 'jjwt' 대신 'auth0/java-jwt' 라이브러리를 사용

RaZel713 committed:
refactor: find와 get의 관습에 따른 메서드 네이밍 변경
